table/block_based_table_reader.cc

```diff
@@ -2491,18 +2494,28 @@ void BlockBasedTableIterator<TBlockIter, TValue>::InitDataBlock() {
 
 template <class TBlockIter, typename TValue>
 void BlockBasedTableIterator<TBlockIter, TValue>::FindKeyForward() {
-  assert(!is_out_of_bound_);
   // TODO the while loop inherits from two-level-iterator. We don't know
   // whether a block can be empty so it can be replaced by an "if".
   while (!block_iter_.Valid()) {
     if (!block_iter_.status().ok()) {
       return;
     }
+    if (read_options_.iterate_upper_bound != nullptr &&
+        block_iter_points_to_real_block_) {
+      Slice index_user_key = index_iter_->key();
+      if (index_key_includes_seq_) {
+        index_user_key = ExtractUserKey(index_user_key);
+      }
+      is_out_of_bound_ =
```

Comment:
There's another issue here. We should not set `is_out_of_bound_` here, because index key of the last block in the file can be larger than the smallest key of the next file in the level. In this case we should not skip the next file when index key is out of bound.

Reply:
Fixed and verified.